core/src/compile/mod.rs

```diff
@@ -507,460 +48,210 @@ impl<'a> NamespaceCompiler<'a> {
+                    }
 
-            let next = self.state.project(next_scope.clone())?;
-            let scope = &mut next.scope;
-            let compiled = &mut next.compiled;
-            let refs = &mut next.refs;
+                    // Then pushbacks to first child of the common root
+                    let (mut root, mut relative_to) = target.relativize(&address);
+                    if !relative_to.is_root() {
+                        let child = relative_to.deeper().unwrap();
+                        root.at(&child);
+                        let is_built = self.state
+                            .project(root.clone())?
+                            .output()
+                            .is_built();
+                        if !is_built && !relative_to.is_root() {
+                            eprintln!("pushback dependency to `{}`", root);
+                            pushbacks.insert(root);
+                        }
+                    }
+                }
 
-            if compiled.is_built() {
-                stage_2!("{}: already built", next_scope);
-                continue;
-            }
+                // Finally, we push direct descendents
+                let descents: BTreeSet<Address> = state
+                    .scope()
+                    .iter_keys()
+                    .filter_map(|child| {
+                        eprintln!("  {}", child);
+                        let is_built = state.get(&child).unwrap().output().is_built();
+                        let child_address = address.clone().into_at(&child);
+                        if !is_built
+                            && !targets.contains(&child_address)
+                            && !pushbacks.contains(&child_address) {
+                            Some(child_address)
+                        } else {
+                            None
+                        }
+                    })
+                    .collect();
+
+                targets
+                    .into_iter()
+                    .chain(pushbacks.into_iter())
+                    .chain(descents.into_iter())
+                    .collect::<Vec<_>>()
+            };
 
-            let own_dependencies: Vec<Scope> = scope
-                .ordering
-                .iter()
-                .filter(|attr| !scope.children.get(*attr).unwrap().compiled.is_built())
-                .map(|attr| next_scope.clone().into_at(attr))
-                .collect();
-            stage_2!(
-                "{}: {} dependencies left",
-                next_scope,
-                own_dependencies.len()
-            );
-
-            let mut foreign_dependencies = Vec::new();
-            for reference in refs.iter() {
-                if !self.vtable.is_built(reference, &next_scope)? {
-                    foreign_dependencies.push(reference.clone());
+            if !next.is_empty() {
+                eprintln!("dependencies not satisfied: {:?}", next);
```

Comment:
Perhaps add a comment here about why this is a cycle condition. Explain that by definition a cycle involves a node being a dependency of itself.